x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/SearchHitCursor.java

```diff
@@ -0,0 +1,224 @@
+        return limit;
+    }
+
+    boolean includeFrozen() {
+        return includeFrozen;
+    }
+
+    @Override
+    public void nextPage(SqlConfiguration cfg, Client client, NamedWriteableRegistry registry, ActionListener<Page> listener) {
+        SearchSourceBuilder q;
+        try {
+            q = Querier.deserializeQuery(registry, nextQuery);
+        } catch (Exception ex) {
+            listener.onFailure(ex);
+            return;
+        }
+
+        SearchSourceBuilder query = q;
+        if (log.isTraceEnabled()) {
+            log.trace("About to execute composite query {}", StringUtils.toString(query));
+        }
+
+        SearchRequest request = Querier.prepareRequest(query, cfg.requestTimeout(), includeFrozen);
+
+        client.search(
+            request,
+            ActionListener.wrap(
+                (SearchResponse response) -> handle(
+                    client,
+                    response,
+                    request.source(),
+                    makeRowSet(query.size(), response),
+                    listener,
+                    includeFrozen
+                ),
+                listener::onFailure
+            )
+        );
+    }
+
+    protected Supplier<SearchHitRowSet> makeRowSet(int sizeRequested, SearchResponse response) {
+        return () -> new SearchHitRowSet(extractors, mask, sizeRequested, limit, response);
+    }
+
+    static void handle(
+        Client client,
+        SearchResponse response,
+        SearchSourceBuilder source,
+        Supplier<SearchHitRowSet> makeRowSet,
+        ActionListener<Page> listener,
+        boolean includeFrozen
+    ) {
+
+        if (log.isTraceEnabled()) {
+            Querier.logSearchResponse(response, log);
+        }
+
+        SearchHit[] hits = response.getHits().getHits();
+
+        SearchHitRowSet rowSet = makeRowSet.get();
+
+        if (rowSet.hasRemaining() == false) {
+            Querier.closePointInTime(
+                client,
+                response.pointInTimeId(),
+                ActionListener.wrap(r -> listener.onResponse(Page.last(rowSet)), listener::onFailure)
+            );
+        } else {
+            source.pointInTimeBuilder(new PointInTimeBuilder(response.pointInTimeId()));
+            updateSearchAfter(hits, source);
+
+            byte[] nextQuery;
+            try {
+                nextQuery = Querier.serializeQuery(source);
+            } catch (IOException e) {
+                listener.onFailure(e);
+                return;
+            }
+
+            SearchHitCursor nextCursor = new SearchHitCursor(
+                nextQuery,
+                rowSet.extractors(),
+                rowSet.mask(),
+                rowSet.getRemainingLimit(),
+                includeFrozen
+            );
+            listener.onResponse(new Page(rowSet, nextCursor));
+        }
+    }
+
+    private static void updateSearchAfter(SearchHit[] hits, SearchSourceBuilder source) {
+        assert hits.length > 0;
```

Comment:
Good point, AIOBE is actually preferable here because it does not tear down the whole world with it. Since this method only gets called if `hasRemaining` is `true` (and that's never the case with empty `hits`) this cannot happen anyway.